File: examples/python/customer/handlers/redeem_loyalty_points.py
"""Handler for RedeemLoyaltyPoints command."""

from datetime import datetime, timezone

import structlog
from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains

from handlers import CommandRejectedError


def handle_redeem_loyalty_points(
    command_book: angzarr.CommandBook,
    command_any: Any,
    state: domains.CustomerState,
    seq: int,
    log: structlog.BoundLogger,
) -> angzarr.EventBook:
    """Handle RedeemLoyaltyPoints command."""
    if not state.name:
        raise CommandRejectedError("Customer does not exist")

    cmd = domains.RedeemLoyaltyPoints()
    command_any.Unpack(cmd)

    if cmd.points <= 0:
        raise CommandRejectedError("Points must be positive")
    if cmd.points > state.loyalty_points:
        raise CommandRejectedError(
            f"Insufficient points: have {state.loyalty_points}, need {cmd.points}"
        )

    new_balance = state.loyalty_points - cmd.points

    log.info(
        "redeeming_loyalty_points",
        points=cmd.points,
        new_balance=new_balance,
        redemption_type=cmd.redemption_type,
    )

    event = domains.LoyaltyPointsRedeemed(
        points=cmd.points,
        new_balance=new_balance,
        redemption_type=cmd.redemption_type,
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[
            angzarr.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )

File: examples/rust/customer/src/handlers/redeem_loyalty_points.rs
//! Handler for RedeemLoyaltyPoints command.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{CustomerState, LoyaltyPointsRedeemed, RedeemLoyaltyPoints};
use common::{decode_command, require_exists, require_positive, BusinessError, Result};

use crate::errmsg;
use crate::state::build_event_response;

/// Handle the RedeemLoyaltyPoints command.
///
/// Redeems loyalty points from an existing customer's balance.
/// Fails if the customer does not exist, points are not positive,
/// or the customer has insufficient points.
pub fn handle_redeem_loyalty_points(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &CustomerState,
    next_seq: u32,
) -> Result<EventBook> {
    require_exists(&state.name, errmsg::CUSTOMER_NOT_FOUND)?;

    let cmd: RedeemLoyaltyPoints = decode_command(command_data)?;

    require_positive(cmd.points, errmsg::POINTS_POSITIVE)?;
    if cmd.points > state.loyalty_points {
        return Err(BusinessError::Rejected(format!(
            "{}: have {}, need {}",
            errmsg::INSUFFICIENT_POINTS,
            state.loyalty_points,
            cmd.points
        )));
    }

    let new_balance = state.loyalty_points - cmd.points;

    let event = LoyaltyPointsRedeemed {
        points: cmd.points,
        new_balance,
        redemption_type: cmd.redemption_type,
    };

    Ok(build_event_response(
        state,
        command_book.cover.clone(),
        next_seq,
        "type.examples/examples.LoyaltyPointsRedeemed",
        event,
    ))
}
